[Very critical fix] Processor>>#terminateRealActive

dionisiydk opened the pull request:
PR is a general fix against the image hanging due to the buggy code simulation in the debugger (during stepping). It should prevent the most critical kind of infinite recursion where it is impossible to recover with cmd+.

PR introduces two methods:
- Processor>>#terminateRealActive
- Processor>>realActiveProcess

The first method here is to terminate the active process that is **really** currently running despite on any effective process installed. 	
It is very critical to ignore the simulation trick of the debugger (the effective process) during normal completion of the process. This method ensures that every process is able to complete. 

The second method here is to suspend the **real** active process during the debugger request triggered from the unhandled error. 
When unhandled error is encountered the execution must be stopped. To stop the execution system needs to find an active process to suspend it. Therefore it is crucial to find the **real** running process. Otherwise the suspend will not stop the execution (it will suspend a wrong process). In most cases the return from the error signal will end up with infinite recursion triggering new debuggers at every step. 

Related changes here:
- Simulation over Process>>#suspend primitive just returns the receiver. We never need to perform real suspend during stepping because we either step over suspended process (a debugger precondition) or we are stepping the wrong process and it would hang the simulator (debugger UI process).
- Context>>#isDead takes into account isEndOfProcessTermination. Now we will consider the context as dead when it is the last completed context in the process
- Three tests to perform full simulation of process execution from the first instruction up to the final "self terminate" instruction of the process. They should be important invariants for any simulation changes:
  - using StepInto
  - using StepOver
  - using StepThrough

dionisiydk committed:
It introduces Processor>>#terminateRealActive to terminates the active process that is really currently running despite on any effective process installed. 	
It is very critical to ignore the simulation trick of the debugger (the effective process) during normal completion of the process. This method ensures that every process is able to complete. 
See the method comment

dionisiydk commented on src/Kernel-Tests-Extended/ProcessTest.class.st:
```diff
@@ -389,6 +389,39 @@ ProcessTest >> testNewProcessWith [
 ]
 
+{ #category : #tests }
+ProcessTest >> testNormalProcessCompletionWithLeftEffectiveProcess [
```
This is a test which reproduces the image hanging. With PR test should pass.

dionisiydk commented on src/Kernel/BlockClosure.class.st:
```diff
@@ -510,6 +512,40 @@ BlockClosure >> on: exception fork: handlerAction [
 ]
 
+{ #category : #exceptions }
+BlockClosure >> on: exception fork: handlerAction return: answerAction [
```
This method was packaged in UnifiedFFI. I think it should be closer to no #return: version.

dionisiydk commented on src/Kernel/ProcessorScheduler.class.st:
```diff
@@ -313,6 +313,31 @@ ProcessorScheduler >> terminateActive [
 ]
 
+{ #category : #'process state change' }
+ProcessorScheduler >> terminateRealActive [
```
Read this comment for more details about the problem

dionisiydk committed:
Introduce Processor>>#realActiveProcess to be used in debugger requests.
Fix  terminating state of normally completed process by moving the flag logic to final termination methods.

dionisiydk commented:
SindarinDebugger and its tests include wrong assumption about process being terminated:
```Smalltalk
SindarinDebugger>>isExecutionFinished
	"Returns whether the debugged execution is finished"
	^ process isTerminating
```
Sindarin uses #terminating flag of the process as the process termination indicator. But it is wrong. 
There is explicit method #isTerminated which should be used there. 
@StevenCostiou Do you have a particular reason to use #terminating flag here?

StevenCostiou commented:
I do not know, this part was done by @dupriezt 
Changing to `#isTerminated` makes Sindarin tests failing, so I think it comes from there: either the sindarin debugger does not configure its debugging contexts properly, or the test miss something in their configuration.
Maybe @dupriezt can give more context?

dionisiydk commented:
> I do not know, this part was done by @dupriezt
> Changing to `#isTerminated` makes Sindarin tests failing, so I think it comes from there: either the sindarin debugger does not configure its debugging contexts properly, or the test miss something in their configuration.

Tests are failing because they put particular expectation on the termination context where debugger is supposed to detect the end of execution. Tests perform the stepping up to the few steps inside #terminateActive until the terminating is set. But the process is considered terminated when it is returned from the #value, not when it is performing #terminateActive. You can check it in workspace: 
```Smalltalk
p := Process
  forContext: 
     [self value.
     isTerminated := Processor activeProcess isTerminated.
     Processor terminateActive] 
  priority: Processor activePriority
p resume.
```
So the debugger should follow the #isTerminated logic. And tests needs to be fixed accordingly

dionisiydk commented:
I see the problem with isTerminated. The first stepInto operation in new process does not step into #value of target block. The context pc in that case is not the initialPC and the process think it corresponds to the return state. Somewhere in these places we need improvements

Ducasse commented:
Thanks you all. 
This is super important that we get this right.

dionisiydk committed:
Change the implementation of #isTerminated method. Now it uses very precise criteria when suspendedContext points to the endPC of Process>>doTerminatinOfYourself method.
The last instruction there is final suspend operation which removes the process from the scheduler.
It means that the process is not terminated until it is really not running (when suspend is executed).
Helper methods are added to support safe stepping (simulation) in the debugger:
- Process>>canBeStepped
- Context>>canBeSteppedSafelly
For the moment #canBeSteppedSafelly is only about preventing stepping over "Processor terminateRealActive". But it can be extended if we will find other cases.
Now the debugger should use these methods as a criteria for the finished execution.

dionisiydk committed:
Missing #isTerminate

dionisiydk committed:
Fix for Sindarin using new method #canBeStepped as a criteria for the finished execution.

dionisiydk committed:
Implement expilit logic to simulate step into #terminateRealActive method. Now the step always delegates the simulation into #terminateActive method to affect the effective process instead of the process running the simulation. It solves all simulation tests (discovered before this commit).
Sindarin changes are reverted back to #terminating flag to prevent stepping over final #suspend. It should be also supported but it's a different story

dionisiydk committed:
Fix isTermination test to reflect #terminateActive delegation

dionisiydk committed:
Better comments

dionisiydk committed:
better #terminateRealActive comment

dionisiydk committed:
Final critical fix: use #realActiveProcess for the debugger requests which are triggered from the process itself as a result of unhandled exception.

Ducasse commented:
Hi denis 
the guys will come back to this issue and give you feedback. 
S.

dionisiydk commented:
> Hi denis
> the guys will come back to this issue and give you feedback.
> S.

Hi @Ducasse . It is very interesting issue. Maybe I will find time to write a blog post about it.

dionisiydk commented on src/Debugger-Oups/OupsDebugRequest.class.st:
```diff
@@ -55,8 +55,15 @@ OupsDebugRequest class >> newForContext: aContext [
+	^ self basicNew 
+			exception: anException;
+			process: Processor realActiveProcess;
+			yourself
```
I will leave @StevenCostiou to refactor it later if needed. 
My suggestion is to make the process retrieval an explicit for the users and remove the lazy initialization.

dionisiydk commented on src/Debugger-Oups/OupsDebugRequest.class.st:
```diff
@@ -55,8 +55,15 @@ OupsDebugRequest class >> newForContext: aContext [
 
 { #category : #'instance creation' }
 OupsDebugRequest class >> newForException: anException [
```
Please sync this changes with actual repository if it exists. I did not find a separate project for Debugger-Oops.

dionisiydk commented on src/Sindarin-Tests/SindarinDebuggerTest.class.st:
```diff
@@ -362,15 +362,16 @@ SindarinDebuggerTest >> testIsExecutionFinished [
-		stepOver;
-		step.
-	"Reached `self isActiveProcess` in Process>>#terminate"
+		stepOver.
```
Due to the changes in termination method the number of steps here were fixed

dionisiydk commented on src/Sindarin-Tests/SindarinDebuggerTest.class.st:
```diff
@@ -362,15 +362,16 @@ SindarinDebuggerTest >> testIsExecutionFinished [
-	"Reached `self isActiveProcess` in Process>>#terminate"
+		stepOver.
+	"Reached `Processor terminateRealActive` method.
+	It is a special termination method which is considered as the end for simulation and should not be stepped" 
```
Please sync this changes with actual repository if it exists. I did not find a separate project for Sindarin

dionisiydk commented:
Tests are unrelated.
And PR is ready for merge

tesonep reviewed (approved):
It looks sound. I will check to rerun the tests because I don't like the error in the FFI marshalling. But if everything goes fine (= this error disappear  I will integrate)

dionisiydk committed:
The final change to use isTerminated in Sindarin to detect #isExecutionFinished.
The actual problem was on the level of kernel. The simulation of the process up to the self termination was not working  using stepInto, stepOver and stepThrough methods.
Corresponding tests are added to cover these basic scenarious. And they are fixed.
The important change is on the primitive simulation: the process #suspend must do nothing during primitive (see the comment). It makes safe the debugging over #suspend call for real active process.

dionisiydk commented on src/Kernel-Tests-Extended/ProcessTest.class.st:
```diff
@@ -7,6 +7,36 @@ Class {
+ProcessTest >> setUp [ 
+	super setUp.
+	
+	self executionProcessMonitor disable
```
Process monitoring of TestExecutionEnvironment should be disabled because tests here are about methods involving into monitoring logic.

dionisiydk commented on src/Kernel/Context.class.st:
```diff
@@ -629,6 +629,14 @@ Context >> doPrimitive: primitiveIndex method: meth receiver: aReceiver args: ar
 			ifFalse: [ ^self send: (arguments at: 2) to: (arguments at: 1) with: (arguments at: 3) lookupIn: (arguments at: 4) ]
 		].
 
+	primitiveIndex = 88 ifTrue: ["Process>>suspend"
```
This is another important change to make debugging safe for stepping over #suspend method (there are cases when suspend is executed over real active process which is an actual UI process of the debugger).

dionisiydk commented on src/Sindarin-Tests/SindarinDebuggerTest.class.st:
```diff
@@ -362,16 +362,10 @@ SindarinDebuggerTest >> testIsExecutionFinished [
+	
+	[ scdbg isExecutionFinished ] whileFalse: [ scdbg stepOver ].
+
+	self assert: scdbg currentProcess isTerminated
```
This change makes test independent from the actual process stack during the self termination.

@StevenCostiou please sync these changes back to actual repository if it exists

dionisiydk commented:
> It looks sound. I will check to rerun the tests because I don't like the error in the FFI marshalling. But if everything goes fine (= this error disappear I will integrate)

Hi @tesonep . Please review my final commit. It is the end of story here.

dionisiydk commented:
> > It looks sound. I will check to rerun the tests because I don't like the error in the FFI marshalling. But if everything goes fine (= this error disappear I will integrate)
> 
> Hi @tesonep . Please review my final commit. It is the end of story here.

One related test is failing. I am checking

isCzech commented:
> I see the problem with isTerminated. [...]

Guys, I've updated #2978 with a suggested fix to #isTerminated. Wondering if it might impact this as well.... Currently a lot of bottom context processes were wrongly reported #isTerminated as terminated.

isCzech commented:
> > I do not know, this part was done by @dupriezt
> > Changing to `#isTerminated` makes Sindarin tests failing, so I think it comes from there: either the sindarin debugger does not configure its debugging contexts properly, or the test miss something in their configuration.
> 
> Tests are failing because they put particular expectation on the termination context where debugger is supposed to detect the end of execution. Tests perform the stepping up to the few steps inside #terminateActive until the terminating is set. But the process is considered terminated when it is returned from the #value, not when it is performing #terminateActive. You can check it in workspace:
> 
> ```smalltalk
> p := Process
>   forContext: 
>      [self value.
>      isTerminated := Processor activeProcess isTerminated.
>      Processor terminateActive] 
>   priority: Processor activePriority
> p resume.
> ```
> 
> So the debugger should follow the #isTerminated logic. And tests needs to be fixed accordingly

Inspired by your example I opened an issue #8656 about creating processes like this one - when you provide a closure instead of a context to #forContext:priority: it may cause a series of consequences for the image and potentially crash it. I'm not sure if your example just contains a mistake or was meant to be run as it is... but an issue is there regardless. Again, I was wondering it this issue might be related to your one.

dionisiydk committed:
Use different criteria to stop the stepping during StepThrough: check that current context is actually the end of process termination.
Using #hasSender: is not reliable criteria for stepping because simulated code can perform some context manipulations breaking this condition in the middle of logic while continuing the execution can still return to the home

dionisiydk committed:
Add #isEndOfProcessTermination check  into #isDead method.
It removes the duplication of checking both criterias. And  it really makes sense in other cases where only #isDead was used. Possibly it wil fix some corner cases during debugger operations over #terminateFromYourself (not normaly accessed by users)

dionisiydk commented on src/Kernel/Context.class.st:
```diff
@@ -973,9 +981,23 @@ Context >> isContext [
+	- it is at the last instruction of process termination (it is unwinded and process is suspended)"
 
-	^ pc isNil
+	^ pc isNil or: [  self isEndOfProcessTermination ]
```
And another important change. 
I checked users of #isDead. The critical senders were already accompanied with #isEndOfProcessTermination as part of this PR to fix existing problems. For the rest of users it also makes sense to take into account this condition. And probably it even fixes some bugs.

dionisiydk committed:
Improve stepOver stop loggic in sindarin debugger: if any of contexts is  dead - stop

dionisiydk commented:
And finally green.
I completed changes from my side. It's ready for a final review and merge

tesonep commented:
Cool, I will check it

isCzech commented:
Hi, how will the new #isTerminated deal with a process that is at the last instruction of the bottom context? Like e.g. 
```p := Process forContext: [Processor activeProcess suspend] asContext  priority: 40. p resume```
Its context isn't dead but it's effectively done working, thus could be considered terminated; resuming it would raise an exception... Or am I missing something? Thanks, Jaromir

dionisiydk commented:
> Hi, how will the new #isTerminated deal with a process that is at the last instruction of the bottom context? Like e.g.
> `p := Process forContext: [Processor activeProcess suspend] asContext priority: 40. p resume`
> Its context isn't dead but it's effectively done working, thus could be considered terminated; resuming it would raise an exception... Or am I missing something? Thanks, Jaromir

Hi @isCzech .

Such process will not be considered as terminated. And after resume it will really continue execution:
- it will return to nil sender
- vm will send #cannotReturn:
- #cannotReturn: will raise an error

The problems which I investigated and fixed here requires to have very concrete "end of process" to fix simulation/debugging of the execution up to the last instruction. Therefore PR introduces little restriction on the form of the process.
As everything it can be improved but it can be done later. We would need to avoid hardcoded #doTerminationFromYourself as a criteria for the "end process method". 

Do you have a **real** use case for #forContext: method which will be broken with my change?

isCzech commented:
Hi, @dionisiydk 

> > Hi, how will the new #isTerminated deal with a process that is at the last instruction of the bottom context? Like e.g.
> > `p := Process forContext: [Processor activeProcess suspend] asContext priority: 40. p resume`
> > Its context isn't dead but it's effectively done working, thus could be considered terminated; resuming it would raise an exception... Or am I missing something? Thanks, Jaromir
> 
> Hi @isCzech .
> 
> Such process will not be considered as terminated.

Thanks, that's an interesting point; after all, terminated processes are the ones you declare terminated, it's not a given, it's a matter of a definition :) In my mind processes at their last instruction were terminated by default even without an explicit termination but maybe it just adds an unnecessary complexity...

> 
> * it will return to nil sender
> * vm will send #cannotReturn:
> * #cannotReturn: will raise an error
> 
> The problems which I investigated and fixed here requires to have very concrete "end of process" to fix simulation/debugging of the execution up to the last instruction. Therefore PR introduces little restriction on the form of the process.
> As everything it can be improved but it can be done later. 

A cosmetic change I'd suggest:
```
doTerminationFromYourself 
	"Stop this process forever from the process itself. 
	Unwind to execute pending ensure:/ifCurtailed: blocks before terminating.
	It assumes that self is the active process. "		
	thisContext unwindForTermination.
>>>	^self suspend    "add a return to remove the second last bytecode instruction"
```
which would simplify your final condition in `Context >> isEndOfProcessTermination` to:
	`^pc = self endPC`
	instead of `^pc = (self endPC - 1)`


> We would need to avoid hardcoded #doTerminationFromYourself as a criteria for the "end process method".
> 

I've had an idea how to avoid hardcoding #doTerminationFromYourself to your termination criteria: you can add a "trap" method, let's call it Process>>#terminated, and forward **all** terminated processes to it. Your #doTerminationFromYourself would send `self terminated` rather than `self suspend`:
```
doTerminationFromYourself 
	"Stop this process forever from the process itself. 
	Unwind to execute pending ensure:/ifCurtailed: blocks before terminating.
	It assumes that self is the active process. "		
	thisContext unwindForTermination.
>>>	^self terminated

terminated
	thisContext terminateTo: nil.   "set thisContext sender to nil"
	self suspend.
	^thisContext restart   "restart and suspend again in case of sending resume to a terminated process"
```
Although this approach doesn't avoid hardcoding a method selector as a criterion, it at least avoids hardcoding a reference to a bytecode position ;) 

With a bit of effort even #doTerminationFromAnotherProcess could forward terminated processes to the #terminated trap, unifying the two termination paths:
```
doTerminationFromAnotherProcess 
	...
	"replace the bottom line with:"
	(suspendedContext notNil and: [ suspendedContext isBottomContext ]) ifTrue: [
>>>		suspendedContext setSender: nil receiver: self method: (Process>>#terminated) arguments: {}]

isTerminated
	self isActiveProcess ifTrue: [ ^ false ].
	
	^ suspendedContext isNil or: [
		suspendedContext isDead or: [
			suspendedContext methodClass == Process and: [
				suspendedContext selector == #terminated ]]]
```
At the first glance all tests are green... 

> Do you have a **real** use case for #forContext: method which will be broken with my change?

No, except maybe if someone wanted to avoid #terminate because it automatically involves unwinds, but other than that, not really. Your slight change in termination criteria just raised my curiosity :)
best,
Jaromir

dionisiydk commented:
Hi @isCzech .

Thanks for your review.
#terminated trap is definitely interesting idea. I would use some fancy name for this like tombstone. 
But I don't want to push it here. It is already too many changes in current logic. So feel free to push it in another PR.

isCzech commented:
> Hi @isCzech .
> 
> Thanks for your review.
> #terminated trap is definitely interesting idea. I would use some fancy name for this like tombstone.
> But I don't want to push it here. It is already too many changes in current logic. So feel free to push it in another PR.

> We would need to avoid hardcoded #doTerminationFromYourself as a criteria for the "end process method".

Hi @dionisiydk ,

Thanks, I sent a PR #8845 as you suggested... Resisted the graveyard inspiration and changed the name of the trap to "#endProcess" ;)

isCzech commented:
Hi @dionisiydk ,
I've opened an issue #8992 (Unwind mechanism during termination is broken - skips multiple unwind blocks and corrupts the image) which addresses a problem that may be related to your issue. I wonder what impact the suggested fix of #8992 could have on this. It only touches #doTerminationFromAnotherProcess and leaves #isTerminated as (broken as) it is. It may surely be considered for merging (if at all) after your PR but still I thought you should know about the termination/unwind issues described in #8992 and PR #8993.
Thanks,
Jaromir

estebanlm commented:
same as with  #8993, we want this, but is very dangerous so I am deferring.

dionisiydk commented:
@tesonep could you please change the source branch to realActiveProcessMerged10 ? I do not have permissions for this.

dionisiydk commented:
I don't understand why CI merges the PR differently than github shows the changes.
For example #testIsExecutionFinished method is original one in the PR image from CI but here you can see that it is changed. 
The latest CI run was using a version of my code merged into Pharo10 branch. The result still the same: wrong merge.
@tesonep do you have any idea what can be wrong? The problem is appears after you switched to Pharo10 branch.

BTW I noticed that CI images are still with Pharo9.0 prefix in the names.

dionisiydk commented:
broken tests are still unrelated

dionisiydk committed:
recover self return in termination method which became lost after merge

dionisiydk commented:
So I found what is the issue. 
The confused merge was only for Sindarin packages. 
In previous Pharo versions they were copies from the official repo (Pharo-Spec/ScriptableDebugger). So that they we not downloaded for every image build. 
In Pharo 10 (maybe the end of Pharo 9) it was changed. Now every build downloads dependencies.

The confusing part is that you can't see this problem. The packages were not removed from this repo. And therefore it's impossible to notice that any code changes will be lost. The diff normally shows only my modifications. 
If packages would be properly cleaned it will be trivially to see what is going on: many packages will be marked as added or removed.

@estebanlm could you please review what other code is still left as a dead in Pharo repo. So that in future it will not kill the other brains.

dionisiydk commented:
I report the separate issue #9822

dionisiydk commented:
The prerequisite for the integration is https://github.com/pharo-spec/ScriptableDebugger/pull/14
It requires new branch for ScriptableDebugger. It is now only master. Then integration this new version into the image. And only after that this PR can be validated again

tesonep commented:
We are seeing this PR with @StevenCostiou in the following weeks.
Do you have a breaking example? 
Is it just hanging when stepping or an error in the debugger?

dionisiydk commented:
> We are seeing this PR with @StevenCostiou in the following weeks.
> Do you have a breaking example?
> Is it just hanging when stepping or an error in the debugger?

The original issue was here: https://github.com/pharo-project/pharo/issues/8443. Just retry the example from the description.
This PR does not fix it. But the PR addresses a general issue why image freezes completely in such "broken debugger" scenarios (#8443 is a simulation bug). So with this PR the #8443 example is recoverable by cmd+. . 

Generally without this PR it is super hard to do any improvements/fixes to the debugger because any error there breaks the image completely.

dionisiydk commented:
Another example: try to step over a continuation resume. 

Generally just try to think for what code the debugger hooks to interrupt execution would not work (hooks in runUntilErrorOrReturnFrom: and some related methods). In all such cases the debugger leaves the process in inconsistent state where it will **never** stop execution.

Ducasse commented:
Thanks denis. Now every two weeks we sit together and review PRs.
We will have a look. We were also thinking that in Java a process cannot terminate another one (if I remember correctly).

dionisiydk commented:
> Thanks denis. Now every two weeks we sit together and review PRs.
> We will have a look. We were also thinking that in Java a process cannot terminate another one (if I remember correctly).

In Java it is really amazing shit I would say. You can't implement a behaviour which will control the life time of arbitrary processes. There is just no way to reliably kill the process with java API (at least it was the case in my Windows Java 7 time)

Ducasse commented:
A naive and honest question: what are the scenario where you need to have this?

dionisiydk commented:
> A naive and honest question: what are the scenario where you need to have this?

For example to control the time of process execution to protect the system from some hanging tasks which it serves.

isCzech commented:
Hi, I noticed Crash on step over #8443 issue has been mentioned here; I've just sent a PR #10383 to fix the root cause of the issue. Please check the explanation of the behavior.
Thanks,
Jaromir

guillep commented:
Hi all, it's been really difficult to review the batch of issues (#8567 , #8845, #9137, #8993, #8509).
We have spent in the last internal sprints several hours taking a look at them, but it is a reality that those changes are difficult to review, some like #8567 are very big, their effects in the system are difficult to foresee independently and even worse in combination.

@tesonep is correctly (!!) proposing to my ear that we make a single PR joining all of them to reduce the merge noise, let the CI test the combination of all issues. I then propose that we eagerly merge this (meta) PR and be ready to quickly apply hot-fixes as soon as possible or revert it in case of fire.

Thoughts? @dionisiydk @Ducasse @isCzech @estebanlm
